pkg/kubernetes/cluster.go

```diff
@@ -0,0 +1,223 @@
+	agentConfig[tokenKey] = serverConfig[tokenKey]
+	agentConfig[cniKey] = serverConfig[cniKey]
+	agentConfig[serverKey] = serverConfig[serverKey]
+	agentConfig[tlsSANKey] = serverConfig[tlsSANKey]
+
+	// Create the initialiser server config
+	initialiserConfig := map[string]any{}
+	for k, v := range serverConfig {
+		initialiserConfig[k] = v
+	}
+	delete(initialiserConfig, serverKey)
+
+	initialiser := identifyInitialiserNode(kubernetes)
+	if initialiser == "" {
+		return nil, fmt.Errorf("failed to determine cluster initialiser")
+	}
+
+	return &Cluster{
+		Initialiser:       initialiser,
+		InitialiserConfig: initialiserConfig,
+		ServerConfig:      serverConfig,
+		AgentConfig:       agentConfig,
+	}, nil
+}
+
+func parseKubernetesConfig(configFile string) (map[string]any, error) {
+	config := map[string]any{}
+
+	b, err := os.ReadFile(configFile)
+	if err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return nil, fmt.Errorf("reading kubernetes config file '%s': %w", configFile, err)
+		}
+
+		zap.S().Warnf("Kubernetes config file '%s' was not provided", configFile)
+
+		// Use an empty config which will be automatically populated later
+		return config, nil
+	}
+
+	if err = yaml.Unmarshal(b, &config); err != nil {
+		return nil, fmt.Errorf("parsing kubernetes config file '%s': %w", configFile, err)
+	}
+
+	return config, nil
+}
+
+func identifyInitialiserNode(kubernetes *image.Kubernetes) string {
+	for _, node := range kubernetes.Nodes {
+		if node.Initialiser {
+			return node.Hostname
+		}
+	}
+
+	// Use the first server node as an initialiser
+	for _, node := range kubernetes.Nodes {
+		if node.Type == image.KubernetesNodeTypeServer {
+			zap.S().Infof("Using '%s' as the cluster initialiser, as one wasn't explicitly selected", node.Hostname)
+			return node.Hostname
+		}
+	}
+
+	return ""
+}
+
+func setSingleNodeConfigDefaults(kubernetes *image.Kubernetes, config map[string]any) {
+	setClusterCNI(config)
+	if kubernetes.Network.APIVIP != "" {
+		appendClusterTLSSAN(config, kubernetes.Network.APIVIP)
+	}
+	if kubernetes.Network.APIHost != "" {
+		appendClusterTLSSAN(config, kubernetes.Network.APIHost)
+	}
+	delete(config, serverKey)
+}
+
+func setMultiNodeConfigDefaults(kubernetes *image.Kubernetes, config map[string]any) {
+	setClusterCNI(config)
+	setClusterToken(config)
+	setClusterAPIAddress(config, kubernetes.Network.APIVIP)
+	appendClusterTLSSAN(config, kubernetes.Network.APIVIP)
+	if kubernetes.Network.APIHost != "" {
+		appendClusterTLSSAN(config, kubernetes.Network.APIHost)
+	}
+}
+
+func setClusterToken(config map[string]any) {
+	if _, ok := config[tokenKey].(string); ok {
+		return
+	}
+
+	token := "foobar" // TODO: generate
+
+	zap.S().Infof("Generated cluster token: %s", token)
+	config[tokenKey] = token
+}
+
+func setClusterCNI(config map[string]any) {
+	if _, ok := config[cniKey]; ok {
+		return
+	}
+
+	auditMessage := fmt.Sprintf("Kubernetes CNI not explicitly set, defaulting to: %s", cniDefaultValue)
```

Comment:
Nit: The user-facing messages have been closer to full sentences in many places, so we may want to consider:

"The Kubernetes CNI is not explicitly set, defaulting to '%s'."